src/webgpu/web_platform/external_texture/video.spec.ts

```diff
@@ -3,13 +3,150 @@ Tests for external textures from HTMLVideoElement (and other video-type sources?
+    video.loop = true;
+    video.muted = true;
+    video.play();
+    await video.play();
+
+    const colorAttachment = t.device.createTexture({
+      format: kFormat,
+      size: { width: kWidth, height: kHeight, depthOrArrayLayers: 1 },
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.RENDER_ATTACHMENT,
+    });
+
+    const pipeline = t.device.createRenderPipeline({
+      vertex: {
+        module: t.device.createShaderModule({
+          code: `
+            [[stage(vertex)]] fn main([[builtin(vertex_index)]] VertexIndex : u32) -> [[builtin(position)]] vec4<f32> {
+                var pos = array<vec4<f32>, 6>(
+                  vec4<f32>( 1.0,  1.0, 0.0, 1.0),
+                  vec4<f32>( 1.0, -1.0, 0.0, 1.0),
+                  vec4<f32>(-1.0, -1.0, 0.0, 1.0),
+                  vec4<f32>( 1.0,  1.0, 0.0, 1.0),
+                  vec4<f32>(-1.0, -1.0, 0.0, 1.0),
+                  vec4<f32>(-1.0,  1.0, 0.0, 1.0)
+                );
+                return pos[VertexIndex];
+            }
+            `,
+        }),
+        entryPoint: 'main',
+      },
+      fragment: {
+        module: t.device.createShaderModule({
+          code: `
+            [[group(0), binding(0)]] var s : sampler;
+            [[group(0), binding(1)]] var t : texture_external;
+
+            [[stage(fragment)]] fn main([[builtin(position)]] FragCoord : vec4<f32>)
+                                     -> [[location(0)]] vec4<f32> {
+                return textureSampleLevel(t, s, FragCoord.xy / vec2<f32>(16.0, 16.0));
+            }
+            `,
+        }),
+        entryPoint: 'main',
+        targets: [
+          {
+            format: kFormat,
+          },
+        ],
+      },
+      primitive: { topology: 'triangle-list' },
+    });
+
+    const linearSampler = t.device.createSampler();
+
+    const externalTextureDescriptor = { source: video };
+    const externalTexture = t.device.importExternalTexture(externalTextureDescriptor);
+
+    const bindGroup = t.device.createBindGroup({
+      layout: pipeline.getBindGroupLayout(0),
+      entries: [
+        {
+          binding: 0,
+          resource: linearSampler,
+        },
+        {
+          binding: 1,
+          resource: externalTexture,
+        },
+      ],
+    });
+    const commandEncoder = t.device.createCommandEncoder();
+    const passEncoder = commandEncoder.beginRenderPass({
+      colorAttachments: [
+        {
+          view: colorAttachment.createView(),
+          loadValue: { r: 0.0, g: 0.0, b: 0.0, a: 1.0 },
+          storeOp: 'store',
+        },
+      ],
+    });
+    passEncoder.setPipeline(pipeline);
+    passEncoder.setBindGroup(0, bindGroup);
+    passEncoder.draw(6);
+    passEncoder.endPass();
+    t.device.queue.submit([commandEncoder.finish()]);
+
+    // Top left corner should be red.
+    t.expectSinglePixelIn2DTexture(
+      colorAttachment,
+      kFormat,
+      { x: 2, y: 2 },
+      {
+        exp: new Uint8Array([0xff, 0x00, 0x00, 0xff]),
```

Comment:
I hope these are the right colors once color management is taken into account! I don't actually have any clue...

Reply:
At least 3 of these tests pass (not sure about mp4...haven't tried with an mp4-enabled build). I'm actually sampling a few pixels away from the edge, which allows me to sample exactly red/green. Added a comment for this.

Reply:
I'm thinking about color profiles/ICC profiles. I don't really know how they work with videos. I assume these test videos are "untagged" insofar as a video can be, so I think these expectations are correct.

(The comment is good though, thanks)